qtile: install proper session .desktop files for both xorg and wayland

arjan-s opened the pull request:
## Description of changes
This PR modifies the qtile package to install the provided session desktop files for login managers.
That in turn allows us to add the package to `services.displayManager.sessionPackages` when the NixOS module is enabled.
After that, both the Xorg and Wayland sessions will be available from login (display) managers, instead of just one session that's hard coded to a specific backend (via the now removed option `backend`).

Additionally, this fixes the `services.xserver.windowManager.qtile.finalPackage` assignment to use the qtile package instead of building a whole new env.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

arjan-s committed:
qtile: use proper session .desktop files

Sigmanificient commented:
![image](https://github.com/NixOS/nixpkgs/assets/53050011/5e4c1774-d62b-49e3-976a-f5daa6c744a4)

Sigmanificient commented:
```
error: Default graphical session, 'none+qtile', not found.
       Valid names for 'services.displayManager.defaultSession' are:
         qtile
         none+icewm
```

Sigmanificient commented:
build my whole system with your branch, seems to work:
![image](https://github.com/NixOS/nixpkgs/assets/53050011/54ea3a35-1425-4673-b143-55286717ada6)

Sigmanificient commented:
@ofborg build qtile-unwrapped.passthru.tests.qtile

arjan-s commented:
Thanks! This PR changed the session names, that's why the test is failing.
Since these changes work on both our systems, I'll just fix the test.

Another todo I thought of after submitting is marking the now-removed backend option deprecated, so people having that in their configs get a nice message.

Sigmanificient commented:
> Another todo I thought of after submitting is marking the now-removed backend option deprecated, so people having that in their configs get a nice message.

Yes this would be nice

arjan-s commented:
Both done!

arjan-s commented:
Looking at this again, I'm afraid the `configFile` option should be deprecated too. But that could break the way some people configure their system, especially those who don't use home-manager (which can write `~/.config/qtile/config.py`).
We can't easily modify the session files that are now installed by the package, so the 'old' method isn't feasible I think.
I'm hoping others have ideas about this...? Perhaps a good solution would be a link to the provided config in the system-wide config dir ([qtile docs here](https://docs.qtile.org/en/stable/manual/config/index.html#configuration-lookup-order))?

Sigmanificient commented:
it look after into the `.desktop` pr, indead would break the `configFile` option.

Setting to `/etc/xdg` would work, but for users that used previous `configFile` options, it may not be the cleanest especially one multiple users system.

I'm a bit new to options, could the desktop file be wrapped in it's own derivation, with a substitubeInPlace `qtile start` -> `qtile start --config ${cfg.configFile}`?

arjan-s commented:
Done. I believe this is ready to test (again).

arjan-s commented:
Converted to draft, because after testing I found that the logic regarding the wrapper and `systemPackages` isn't right. I'll have to dive deeper to get to a point where it just works.

arjan-s commented:
Okay, I believe I have now (finally) arrived at a working and much cleaner solution!
* I removed the wrapper since we weren't actually using it anyway (we built a separate new env for display managers and we added the unwrapped package to `environment.systemPackages`).
* The `extraPackages` are now nicely added to the dependencies of Qtile itself, making https://github.com/NixOS/nixpkgs/pull/299843 a thing of the past (I hope).
* The `finalPackage` (which optionally contains `qtile-extras` and other dependencies) is used for both display managers and added to $PATH (the latter makes commands like `qtile check` work).
* If a `configFile` is provided, it is symlinked to `/etc/xdg/qtile/config.py`, which is automatically used by Qtile if `~/.config/qtile/config.py` isn't found.

drupol reviewed (approved):
LGTM

SuperSandro2000 commented on nixos/modules/services/x11/window-managers/qtile.nix:
```diff
@@ -26,7 +26,7 @@ in
       default = "x11";
       description = ''
-          Backend to use in qtile: `x11` or `wayland`.
+          This option is deprecated. The qtile package now provides separate display sessions for both X11 and Wayland.
```
We should add an assetion or probably even better use mkRemovedOption

arjan-s replied:
Much better indeed, fixed that now.

jwijenbergh reviewed (approved):
Thanks a lot for making these changes. Looks good to me. Should we maybe move qtile later to `programs.qtile`? Most wayland compositors do that, but I'm not sure for qtile as it also supports x11

acid-bong commented:
Should've made a warning for `qtile` instead of removing it tho 😕

Nebucatnetzer commented:
> Should've made a warning for `qtile` instead of removing it tho 😕

Yeah was quite surprised today when I tested the update to 24.11.
